File: python3/valid_parentheses.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def isValid(self, s: str) -> bool:
        n = len(s)
        stack = []

        for i in range(n):
            if not stack:
                stack.append(s[i])
                continue

            if s[i] == '(' or s[i] == '[' or s[i] == '{':
                stack.append(s[i])
            else:
                if s[i] == ')' and stack[-1] == '(':
                    stack.pop()
                elif s[i] == ']' and stack[-1] == '[':
                    stack.pop()
                elif s[i] == '}' and stack[-1] == '{':
                    stack.pop()
                else:
                    return False

        if stack:
            return False

        return True

File: golang/valid_parentheses.go
package main

// Space :	O(n)
// Time	 :	O(n)

func isValid(s string) bool {
	stack := []string{}

	for _, c := range s {
		chr := string(c)

		if len(stack) == 0 {
			stack = append(stack, chr)
			continue
		}

		if chr == "(" || chr == "{" || chr == "[" {
			stack = append(stack, chr)
		} else {
			if stack[len(stack)-1] == "(" && chr == ")" {
				stack = stack[:len(stack)-1]
			} else if stack[len(stack)-1] == "{" && chr == "}" {
				stack = stack[:len(stack)-1]
			} else if stack[len(stack)-1] == "[" && chr == "]" {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}

	if len(stack) > 0 {
		return false
	}

	return true
}
